### PR TITLE
[dotnet] Add nullability annotations to print types

### PR DESCRIPTION
**Pull request by RenderMichael**
### **User description**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description

This adds nullability annotations to the PDF-print stack, as well as documenting potential exceptions alongside some code modernization.

### Motivation and Context

Contributes to #14640 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

___

### **PR Type**
Enhancement, Bug fix, Tests

___

### **Description**
- Enabled nullability annotations across multiple classes to improve code safety and clarity.
- Added detailed exception documentation to constructors and methods to specify potential errors.
- Refactored properties to use auto-properties for cleaner code.
- Added validation checks and exception handling for invalid input values in `PrintOptions`.
- Enhanced test coverage by adding tests for invalid page numbers, ranges, and negative values.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>EncodedFile.cs</strong><dd><code>Add nullability and refactor EncodedFile class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/EncodedFile.cs

<li>Enabled nullability annotations.<br> <li> Added exception documentation for constructor.<br> <li> Refactored properties to use auto-properties.<br> <li> Simplified <code>ToString</code> method.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-b02ffe7cb5b5edb6002bd71fb537f8e60d8316f61c9237e75838f7b9765f77d3">+14/-17</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>ISupportsPrint.cs</strong><dd><code>Add nullability and exception documentation to ISupportsPrint</code></dd></summary>
<hr>

dotnet/src/webdriver/ISupportsPrint.cs

<li>Enabled nullability annotations.<br> <li> Added exception documentation for <code>Print</code> method.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-96ff6f89ebc0debefcfdf691c7808746d6de8c45b6e12733b5c4575908004ef5">+7/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>PrintDocument.cs</strong><dd><code>Enhance PrintDocument with nullability and exceptions</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/PrintDocument.cs

<li>Enabled nullability annotations.<br> <li> Added exception documentation for constructor and <code>SaveAsFile</code> method.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-c74c4d0b95fdba3c306ae5e4f4450e052286fd3fd535417a0d3b284c669a3f12">+17/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>PrintOptions.cs</strong><dd><code>Improve PrintOptions with nullability and validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/PrintOptions.cs

<li>Enabled nullability annotations.<br> <li> Added exception handling for invalid values.<br> <li> Refactored properties and methods for clarity.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-b7252358e015a773e2d453e794cf8c614fcd330985e0ddf4fe3202392f508e2c">+89/-79</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Screenshot.cs</strong><dd><code>Add nullability and exceptions to Screenshot class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/Screenshot.cs

<li>Enabled nullability annotations.<br> <li> Added exception documentation for constructor.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-e313f6ac04959ee687e6a0c729b9ef4c1ec53846da4730ea06751d92072c3092">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>WebDriver.cs</strong><dd><code>Add null check in WebDriver Print method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/src/webdriver/WebDriver.cs

- Added null check for `printOptions` in `Print` method.

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-ef1874c09b1dca0d62ea1f0ead685355944b9cb7ae63df60ad9b32abdf90c8dc">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>PrintTest.cs</strong><dd><code>Add validation tests for PrintOptions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

dotnet/test/common/PrintTest.cs

<li>Added tests for invalid page numbers and ranges.<br> <li> Added tests for negative page sizes and margins.<br>

</details>

  </td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/14773/files#diff-e86bfd82d820efb866b31c104ebec828f8ca2c6cc0feee2ff4b90ce42ab937d4">+34/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: dotnet/src/webdriver/ISupportsPrint.cs [hunk 1]
```diff
@@ -27,8 +31,9 @@ public interface ISupportsPrint
         /// <summary>
         /// Gets a <see cref="PrintDocument"/> object representing a PDF-formatted print representation of the page.
         /// </summary>
-        /// <param name="options">A <see cref="PrintOptions"/> object describing the options of the printed document.</param>
+        /// <param name="printOptions">A <see cref="PrintOptions"/> object describing the options of the printed document.</param>
         /// <returns>The <see cref="PrintDocument"/> object containing the PDF-formatted print representation of the page.</returns>
-        PrintDocument Print(PrintOptions options);
+        /// <exception cref="ArgumentNullException">If <paramref name="printOptions"/> is <see langword="null"/>.</exception>
+        PrintDocument Print(PrintOptions printOptions);
```
Comment (RenderMichael) [inline dotnet/src/webdriver/ISupportsPrint.cs]: Technically breaking change here, the `WebDriver.Print` method and the `ISupportsPrint.Print` methods have different parameter names. It's best practice to align them (there's an analyzer we could turn on) and the `ArgumentNullException` thrown will align to the runtime one.

I figure more people are calling this method directly on `WebDriver` rather than this interface, so I aligned the name with that one. However, if someone is calling `supportsPrint.Print(options: myOptions);` then they will be broken (very simple and obvious fix on their end).

File: dotnet/test/common/PrintTest.cs [hunk 2]
```diff
@@ -89,5 +90,38 @@ public void MarginsCannotBeNull()
...
+        }
+
+        [Test]
+        public void PageSizeCannotHaveNegativeValues()
+        {
+            Assert.That(() => new PrintOptions.PageSize { Height = -1 }, Throws.TypeOf<ArgumentOutOfRangeException>());
+            Assert.That(() => new PrintOptions.PageSize { HeightInInches = -1 }, Throws.TypeOf<ArgumentOutOfRangeException>());
+            Assert.That(() => new PrintOptions.PageSize { Width = -1 }, Throws.TypeOf<ArgumentOutOfRangeException>());
+            Assert.That(() => new PrintOptions.PageSize { WidthInInches = -1 }, Throws.TypeOf<ArgumentOutOfRangeException>());
+        }
```
Comment (RenderMichael) [inline dotnet/test/common/PrintTest.cs:115]: Two things here that are potentially bugs:

I tried to add this assertion:
```csharp
Assert.That(() =>
{
    var options = new PrintOptions();
    options.AddPageRangeToPrint("1-2");
    options.AddPageToPrint(1);
}, Throws.TypeOf<ArgumentException>());
```

but no exception is thrown. Validating against this is not trivial, but there's already so much validation on this method that I figured it may be worthwhile.

Second, the following code does nothing:
```csharp
options.AddPageRangeToPrint("1");
```

No exception is thrown, the single page is not added, nothing. I can fix that simply. Should it throw, or accept the single page?

File: dotnet/src/webdriver/PrintOptions.cs [hunk 3]
```diff
@@ -148,77 +134,91 @@ public void AddPageToPrint(int pageNumber)
...
             {
-                throw new ArgumentException("Page range cannot have multiple separators");
+                throw new ArgumentException("Page range cannot have multiple separators", nameof(pageRange));
             }
 
             int startPage = ParsePageRangePart(pageRangeParts[0], 1);
             if (startPage < 1)
             {
-                throw new ArgumentException("Start of a page range must be greater than or equal to 1");
+                throw new ArgumentOutOfRangeException(nameof(pageRange), "Start of a page range must be greater than or equal to 1");
```
Comment (RenderMichael) [inline dotnet/src/webdriver/PrintOptions.cs:172]: Some of these exceptions are better off as `ArgumentOutOfRangeException`s. Luckily this extends `ArgumentException` so there is no breaking change.

File: dotnet/src/webdriver/ISupportsPrint.cs (hunk 1 above)
Comment (nvborisenko) [reply dotnet/src/webdriver/ISupportsPrint.cs]: No, please don't do it here. Probably in separate PR where we will think about it.

File: dotnet/test/common/PrintTest.cs (hunk 2 above)
Comment (nvborisenko) [reply dotnet/test/common/PrintTest.cs:115]: As separate PR please.

File: dotnet/src/webdriver/PrintOptions.cs (hunk 3 above)
Comment (nvborisenko) [reply dotnet/src/webdriver/PrintOptions.cs:172]: It is a breaking change (behavior), but I think we can accept it, even in this PR.

File: dotnet/src/webdriver/ISupportsPrint.cs (hunk 1 above)
Comment (RenderMichael) [reply dotnet/src/webdriver/ISupportsPrint.cs]: Reverted

File: dotnet/test/common/PrintTest.cs (hunk 2 above)
Comment (RenderMichael) [reply dotnet/test/common/PrintTest.cs:115]: Opened https://github.com/SeleniumHQ/selenium/issues/14783

File: dotnet/test/common/PrintTest.cs [hunk 4]
```diff
@@ -72,6 +72,7 @@ public void CanPrintWithMostParams()
             };
 
             options.AddPageRangeToPrint("1-3");
+            options.AddPageToPrint(1);
```
Comment (RenderMichael) [inline dotnet/test/common/PrintTest.cs]:
```suggestion
```

File: dotnet/src/webdriver/WebDriver.cs [hunk 5]
```diff
@@ -378,6 +378,7 @@ public PrintDocument Print(PrintOptions printOptions)
         {
             if (printOptions is null)
             {
+                // TODO the interface names this parameter 'options', but we will be throwing for 'printOptions'. Confusion for users?
```
Comment (nvborisenko) [inline dotnet/src/webdriver/WebDriver.cs]: No please, todo comment in "legacy" will be never resolved. Manage it via issue tracker.

Comment (RenderMichael) [reply dotnet/src/webdriver/WebDriver.cs]: Opened https://github.com/SeleniumHQ/selenium/issues/14784

```suggestion
```

File: dotnet/src/webdriver/EncodedFile.cs [hunk 6]
```diff
@@ -19,41 +19,41 @@
...
+        /// <para>-or-</para>
+        /// <para>The format of <paramref name="base64EncodedFile"/> is invalid. <paramref name="base64EncodedFile"/> contains a non-base-64 character,
+        /// more than two padding characters, or a non-white space-character among the padding characters.</para>
+        /// </exception>
         protected EncodedFile(string base64EncodedFile)
         {
-            this.base64Encoded = base64EncodedFile;
-            this.byteArray = Convert.FromBase64String(this.base64Encoded);
+            this.AsBase64EncodedString = base64EncodedFile ?? throw new ArgumentNullException(nameof(base64EncodedFile));
+            this.AsByteArray = Convert.FromBase64String(this.AsBase64EncodedString);
```
Comment (nvborisenko) [inline dotnet/src/webdriver/EncodedFile.cs]: We can defer computation: when user gets `AsByteArray` property getter, we can compute and return. Safe and minor change, as you wish (ideally as separate PR).

**Review by nvborisenko — APPROVED** [review]

**Comment by nvborisenko** [comment]
Thanks @RenderMichael, safe to merge (if CI is green), after release process of 4.27 is finished.

File: dotnet/src/webdriver/EncodedFile.cs (hunk 6 above)
Comment (nvborisenko) [reply dotnet/src/webdriver/EncodedFile.cs]: Ah no, in this PR please. `Convert.FromBase64String(this.AsBase64EncodedString);` relies on the state (this.AsBase64EncodedString) which might be not computed (just prevent issues in future).

**Review by nvborisenko — CHANGES_REQUESTED** [review]
Please fix one comment.

File: dotnet/src/webdriver/EncodedFile.cs (hunk 6 above)
Comment (RenderMichael) [reply dotnet/src/webdriver/EncodedFile.cs]: I fixed this now to use the local ctor parameter.

However, I'm noticing there's a flaw here, since the array is writeable. Someone could change the value of the array and break things.

I vote to `[Obsolete]` the `AsByteArray` property, and instead add a `ToByteArray()` method which makes a new copy of the array each time. Targetting .NET 8, we can also make an `AsSpan` member which exposes the array as a `ReadOnlySpan<byte>`.

What do we think of this? I shouldn't do it in this PR, but if you think this is worthy of pursuit, I can open an issue for it.

Comment (nvborisenko) [reply dotnet/src/webdriver/EncodedFile.cs]: > I fixed this now to use the local ctor parameter.

Accepted.

> I vote to [Obsolete] the AsByteArray property, and instead add a ToByteArray()

You are right, I will accept it, as separate PR.

> we can also make an AsSpan

Oh not now, it is one more big step to make selenium modern, let's finish with AOT and NRT :D

File: dotnet/src/webdriver/EncodedFile.cs [hunk 7]
```diff
@@ -19,41 +19,41 @@
...
     {
-        private string base64Encoded = string.Empty;
-        private byte[] byteArray;
-
         /// <summary>
         /// Initializes a new instance of the <see cref="EncodedFile"/> class.
         /// </summary>
         /// <param name="base64EncodedFile">The file as a Base64-encoded string.</param>
+        /// <exception cref="ArgumentNullException">If <paramref name="base64EncodedFile"/> is <see langword="null"/>.</exception>
+        /// <exception cref="FormatException">
```
Comment (nvborisenko) [inline dotnet/src/webdriver/EncodedFile.cs:36]: I am not so experienced guy writing docs, but does it mean that we should mention **all** possible exception types, even raised by downstream flow?!

Comment (RenderMichael) [reply dotnet/src/webdriver/EncodedFile.cs:36]: Only because we are exposing the `EncodedFile` constructor to other users! Users can derive this type and do something like this:

```csharp
public MyFile : EncodedFile
{
    public MyFile(string encodedFile) : base(encodedFile)
    {
    }
}
```

 If this is not intended for users to do, we can change this constructor to `private protected` and we do not need to bother documenting it at all!

Comment (nvborisenko) [reply dotnet/src/webdriver/EncodedFile.cs:36]: I googled, seems a good library should document all known exceptions that might be raised by the member, even transitively.

**Review by nvborisenko — APPROVED** [review]

**Comment by RenderMichael** [comment]
Thanks!

I saw the recent 4.27 release; it was nice to see the .NET release notes section longer than previous releases.

**Comment by nvborisenko** [comment]
The changes are not related to failed CI tests (ShadowRoot). Merging it.
